lib/web_ui/lib/src/engine/view_embedder/application_dom/application_dom.dart

```diff
@@ -0,0 +1,95 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:meta/meta.dart';
+
+import '../../dom.dart';
+
+import '../hot_restart_cache_handler.dart';
+import 'custom_element_application_dom.dart';
+import 'full_page_application_dom.dart';
+
+/// Provides the API that the FlutterViewEmbedder uses to interact with the DOM.
+///
+/// The base class handles "global" stuff that is shared across implementations,
+/// like handling hot-restart cleanup.
+///
+/// This class is specialized to handle different types of DOM embeddings:
+///
+/// * [FullPageApplicationDom] - The default behavior, where flutter takes
+///   control of the whole web page. This is how Flutter Web used to operate.
+/// * [CustomElementApplicationDom] - Flutter is rendered inside a custom host
+///   element, provided by the web app programmer through the engine
+///   initialization.
+abstract class ApplicationDom {
+  @mustCallSuper
+  ApplicationDom() {
+    // Prepare some global stuff...
+    assert(() {
+      _hotRestartCache = HotRestartCacheHandler();
+      return true;
+    }());
+  }
+
+  factory ApplicationDom.create({DomElement? hostElement}) {
+    if (hostElement != null) {
+      return CustomElementApplicationDom(hostElement);
+    } else {
+      return FullPageApplicationDom();
+    }
+  }
+
+  /// Keeps a list of elements to be cleaned up at hot-restart.
+  HotRestartCacheHandler? _hotRestartCache;
+
+  /// Returns the 'type' (a String for debugging) of the ApplicationDom instance.
+  String get type;
+
+  /// Sets-up the viewport Meta-Tag for the app.
+  void applyViewportMeta();
+
+  /// Sets the global styles for the hostElement of this Flutter web app.
+  ///
+  /// [font] is the CSS shorthand property to set all the different font properties.
+  void setHostStyles({
+    required String font,
+  });
+
+  /// Sets an attribute in the hostElement.
+  ///
+  /// Like "flt-renderer" or "flt-build-mode".
+  void setHostAttribute(String name, String value);
+
+  /// Attaches the glassPane element into the hostElement.
+  void attachGlassPane(DomElement glassPaneElement);
+
+  /// Attaches the resourceHost element into the hostElement.
+  void attachResourcesHost(DomElement resourceHost, { DomElement? nextTo });
+
+  /// Register a listener for window resize events
+  void setMetricsChangeHandler(void Function(DomEvent? event) handler);
+
+  /// Register a listener for locale change events.
+  void setLanguageChangeHandler(void Function(DomEvent event) handler);
```

Comment:
I think this is way too many methods. Instead, I would just have one: `initialize`. Then each implementation can start with a copy&paste of the original `ViewEmbedder.reset` and customize it to its needs.

Reply:
Removed the metricsChangeHandler to a Stream coming from the DimensionsProvider, exposed through the EngineWindow!

Reply:
I still don't know what to do with the languageChangeHandler. I could listen to mutations to the attribute "lang" of the hostElement, BUUUT I'm not sure anybody will want that behavior :)

Reply:
Language is expose through `SingleFlutterWindow` only. `FlutterView` doesn't have it. So I think this is an inherently global property and should live outside `ApplicationDom`.

https://api.flutter.dev/flutter/dart-ui/SingletonFlutterWindow-class.html
https://api.flutter.dev/flutter/dart-ui/FlutterView-class.html

Reply:
I shoehorned it here because I was attempting to reuse the same `onHotRestart` handler, but since we can have a bunch of those in the engine, it can be self contained and go elsewhere. I'll refactor it elsewhere; and that'll let me simplify the code of the Cache class so it only handles the DOM elements that we want to delete after hot restart.

Reply:
I moved this to the EnginePlatformDispatcher, near other global configuration coming from the browser, like the brightness, high contrast, and font size.

It is also now self contained within the dispatcher. The old implementation was only calling platformDispatcher methods in a very roundabout way.